### PR TITLE
Add domain parameter match check for DH and ECDH key exchange.

### LINKED ISSUE #14808: DH key derivation checks are incomplete (alpha14)
Commit https://github.com/openssl/openssl/commit/e454a3934c287aede194cac49c8934f04bf6a04f#diff-78ed4660fbeb3ac4a81589478f963d94a1452d534c6d7189d562512852a1efcfR79-R94 shifts DH key checks into `EVP_PKEY_derive_set_peer`. This changes the behavior of `DH_compute_key` (which is fine, OpenSSL 3.0 is a breaking change), but the new code path does not do the same set of checks that were previously performed when calling `DH_compute_key`. Specifically, it appears the verification that the public key is less than `p` is missing:

https://github.com/openssl/openssl/blob/dfccfde06562ac87fe5e5f9401ba86cad050d9a2/crypto/ffc/ffc_key_validate.c#L41-L48

This causes a new failure in the pyca/cryptography test suite where we verify that DH kex fails when passing this vector. This vector fails as expected on OpenSSL 1.0.2+ including alpha13, but passes in alpha14.

https://github.com/pyca/cryptography/blob/58f0ad5b6b928677931c7ad44deee839a29ee9d3/vectors/cryptography_vectors/asymmetric/DH/bad_exchange.txt#L3-L11

### PR DESCRIPTION
**Pull request by slontis**
Fixes #14808

Validation checks were moved into EVP_PKEY_derive_set_peer() which broke
an external negative test. Originally the old code was semi working by checking the peers public key was in the range of other parties p. It was not actually ever
checking that the domain parameters were consistent between the 2
parties. It now checks the parameters match as well as validating the
peers public key.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
rebased

**Comment by slontis** [comment]
The appveyor issue is unrelated to this PR (See https://github.com/openssl/openssl/pull/14851)

File: providers/common/provider_err.c [hunk 1]
```diff
@@ -111,6 +111,8 @@ static const ERR_STRING_DATA PROV_str_reasons[] = {
     {ERR_PACK(ERR_LIB_PROV, 0, PROV_R_KEY_SIZE_TOO_SMALL),
     "key size too small"},
     {ERR_PACK(ERR_LIB_PROV, 0, PROV_R_LENGTH_TOO_LARGE), "length too large"},
+    {ERR_PACK(ERR_LIB_PROV, 0, PROV_R_MISMATCHING_SHARED_PARAMETERS),
```
Comment (paulidale) [inline providers/common/provider_err.c:114]: _shared_ or _domain_?  Not sure which is better.

**Review by paulidale — APPROVED** [review]

File: providers/common/provider_err.c (hunk 1 above)
Comment (slontis) [reply providers/common/provider_err.c:114]: Yeah I had that and then changed it - Domain parameters applies to ec also.. Do you want it renamed?

Comment (t8m) [reply providers/common/provider_err.c:114]: We use domain parameters elsewhere. Please do.

**Review by t8m — COMMENTED** [review]
Please rename the error to PROV_R_MISMATCHING_DOMAIN_PARAMETERS or PROV_R_DOMAIN_PARAMETERS_MISMATCH. Otherwise OK.

**Comment by slontis** [comment]
Apart from renaming the error there are no changes. Rebased also.

**Comment by slontis** [comment]
I am assuming the approval still holds and that this is ready for merge if the CI loop passes..

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
pauli - can I merge this now?

**Comment by slontis** [comment]
Thanks for reviewing. Merged to master.

**Pull request closed without merging**